codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -69,409 +71,1060 @@
+                    // Clear early in case fireChildReadComplete() causes it to need to be re-processed
+                    childChannel.fireChannelReadPending = false;
+                    if (childChannel.fireChildReadComplete()) {
+                        flushNeeded = true;
+                    }
+                }
+                childChannel.next = null;
+                current = current.next;
+            }
+        } finally {
+            tail = head = null;
+
+            // We always flush as this is what Http2ConnectionHandler does for now.
+            // TODO: I think this is not really necessary and we should be able to optimize this in the future by
+            // checking flushNeeded and only flush if this returns true.
+            flush0(ctx);
         }
-        channel.pipeline().addLast(handler);
+    }
 
-        initOpts(channel, options);
-        initAttrs(channel, attrs);
+    @Override
+    public final void flush(ChannelHandlerContext ctx) {
+        flushNeeded = false;
+        super.flush(ctx);
+    }
 
-        ChannelFuture future = group.register(channel);
-        // Handle any errors that occurred on the local thread while registering. Even though
-        // failures can happen after this point, they will be handled by the channel by closing the
-        // channel.
-        if (future.cause() != null) {
-            if (channel.isRegistered()) {
-                channel.close();
-            } else {
-                channel.unsafe().closeForcibly();
-            }
-        }
-        return future;
+    // Allow to override for testing
+    void flush0(ChannelHandlerContext ctx) {
+        flush(ctx);
     }
 
-    @SuppressWarnings("unchecked")
-    private static void initOpts(Channel channel, Map<ChannelOption<?>, Object> opts) {
-        if (opts != null) {
-            for (Entry<ChannelOption<?>, Object> e: opts.entrySet()) {
-                try {
-                    if (!channel.config().setOption((ChannelOption<Object>) e.getKey(), e.getValue())) {
-                        logger.warn("Unknown channel option: " + e);
-                    }
-                } catch (Throwable t) {
-                    logger.warn("Failed to set a channel option: " + channel, t);
+    // Allow to override for testing
+    void onBytesConsumed(@SuppressWarnings("unused") ChannelHandlerContext ctx,
+                         Http2FrameStream stream, int bytes) throws Http2Exception {
+        consumeBytes(stream.id(), bytes);
+    }
+
+    // Allow to extend for testing
+    static class Http2MultiplexCodecStream extends DefaultHttp2FrameStream {
+        DefaultHttp2StreamChannel channel;
+    }
+
+    private enum ReadState {
+        READ_QUEUED,
+        READ_IGNORED_CHANNEL_INACTIVE,
+        READ_PROCESSED_BUT_STOP_READING,
+        READ_PROCESSED_OK_TO_PROCESS_MORE
+    }
+
+    // TODO: Handle writability changes due writing from outside the eventloop.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)